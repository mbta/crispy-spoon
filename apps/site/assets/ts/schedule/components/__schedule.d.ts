import { TypedRoutes } from "../../stop/components/__stop";
import { Route } from "../../__v3api";

export interface SchedulePageData {
  connections: TypedRoutes[];
  pdfs: SchedulePDF[];
  teasers: string | null;
  hours: string;
  fares: Fare[];
  fare_link: string;
  holidays: Holiday[];
  route: Route;
  schedule_note: ScheduleNote | null;
  stops: SimpleStop[];
}

export interface ScheduleNote {
  peak_service: string;
  offpeak_service: string;
  exceptions: ServiceException[];
}

export interface ServiceException {
  type: string;
  service: string;
}

export interface SimpleStop {
  id: string;
  name: string;
}

export interface SchedulePDF {
  title: string;
  url: string;
}

export interface Fare {
  title: string;
  price: string;
}

export interface Holiday {
  name: string;
  date: string;
}
