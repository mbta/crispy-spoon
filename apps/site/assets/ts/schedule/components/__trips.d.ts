import { Route, Trip, Stop, PredictedOrScheduledTime } from "../../__v3api";

export interface Journey {
  trip: Trip;
  route: Route;
  departure: Departure;
}

export interface EnhancedJourney extends Journey {
  realtime: PredictedOrScheduledTime;
}

export interface Departure {
  time: string;
  schedule: Schedule | null;
  prediction: Prediction | null;
}

export interface TripDeparture {
  schedule: TripSchedule;
  prediction: Prediction | null;
}

export interface TripInfo {
  times: TripDeparture[];
  vehicle: Vehicle | null;
  vehicle_stop_name: string;
  stop_count: number;
  status: string;
  duration: number;
  fare: Fare;
}

export interface Prediction {
  delay: number;
  status: string | null;
  track: string | null;
}

export interface Schedule {
  time: string;
  stop_sequence: number;
  pickup_type: number;
  "flag?": boolean;
  "early_departure?": boolean;
  "last_stop?": boolean;
}

export interface TripSchedule extends Schedule {
  stop: Stop;
  fare: Fare;
}

export interface Vehicle {
  trip_id: string;
  stop_id: string;
  status: string;
}

export interface Fare {
  price: string;
  fare_link: string;
}
