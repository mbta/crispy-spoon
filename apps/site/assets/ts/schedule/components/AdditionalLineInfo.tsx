import React, { ReactElement } from "react";
import { Holiday, SchedulePDF, Fare, ScheduleNote } from "./__schedule";
import { EnhancedRoute } from "../../__v3api";
import { TypedRoutes } from "../../stop/components/__stop";
import PDFSchedules from "./PDFSchedules";
import Connections from "./Connections";
import ContentTeasers from "./ContentTeasers";
import Fares from "./Fares";
import UpcomingHolidays from "./UpcomingHolidays";
import HoursOfOperation from "./HoursOfOperation";
import { isSubwayRoute } from "../../models/route";

interface Props {
  teasers: string | null;
  pdfs: SchedulePDF[];
  connections: TypedRoutes[];
  fares: Fare[];
  fareLink: string;
  route: EnhancedRoute;
  holidays: Holiday[];
  scheduleNote: ScheduleNote | null;
  hours: string;
}

const AdditionalLineInfo = ({
  connections,
  pdfs,
  teasers,
  fares,
  holidays,
  fareLink,
  route,
  hours,
  scheduleNote
}: Props): ReactElement<HTMLElement> => (
  <>
    <ContentTeasers teasers={teasers} />
    <PDFSchedules pdfs={pdfs} />
    <Connections connections={connections} />
    <Fares fares={fares} fareLink={fareLink} routeType={route.type} />
    {/* Only show the hours for non subway lines */}
    {!isSubwayRoute(route) && (
      <HoursOfOperation
        route={route}
        pdfs={pdfs}
        scheduleNote={scheduleNote}
        hours={hours}
      />
    )}
    <UpcomingHolidays holidays={holidays} />
  </>
);

export default AdditionalLineInfo;
