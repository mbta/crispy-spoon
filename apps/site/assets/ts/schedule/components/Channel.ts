import { Socket, Channel } from "phoenix";

declare global {
  interface Window {
    channels: { [id: string]: Channel };
    socket?: Socket;
  }
}

type UpdateEventName = "reset" | "add" | "update" | "setChannel";

interface UpdateEvent<T> {
  event: UpdateEventName;
  data: T;
}

interface RemoveEvent {
  event: "remove";
  data: string[];
}

export type SocketEvent<T> = UpdateEvent<T> | RemoveEvent;

export const doInitChannel = (id: string): Channel => {
  if (!window.socket) {
    window.socket = new Socket("/socket", {});
    window.socket.connect();
  }

  if (!window.channels) {
    window.channels = {};
  }

  if (window.channels[id]) return window.channels[id];

  const channel = window.socket.channel(id, {});
  window.channels[id] = channel;

  channel
    .join()
    .receive("error", ({ reason }) =>
      // eslint-disable-next-line no-console
      console.error(`failed to join ${id}`, reason)
    )
    // eslint-disable-next-line no-console
    .receive("timeout", () => console.error(`timeout joining ${id}`))
    .receive("ok", () => {
      if (id !== "vehicles:remove") {
        const route = id.split(":");
        channel.push("init", {
          // eslint-disable-next-line camelcase
          route_id: route[1],
          // eslint-disable-next-line camelcase
          direction_id: route[2]
        });
      }
    });

  return channel;
};

export const initChannel = <T>(
  channelId: string,
  onData: (data: SocketEvent<T>) => void
): void => {
  const channel = doInitChannel(channelId);
  channel.on("data", onData);
};

// NB: this removes *all* handlers from the channel. We're currently throwing
// away the reference returned by `channel.on`, so we can't remove a specific
// handler, but we don't have a use case for that yet.
export const stopChannel = (id: string): void => {
  if (window.channels && window.channels[id]) {
    window.channels[id].off("data");
  }
};
