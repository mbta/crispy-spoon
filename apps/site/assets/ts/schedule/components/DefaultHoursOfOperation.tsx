import React, { ReactElement } from "react";
import ExpandableBlock from "../../components/ExpandableBlock";

const DefaultHoursOfOperation = ({
  hours
}: {
  hours: string;
}): ReactElement<HTMLElement> | null =>
  hours ? (
    <ExpandableBlock
      header={{ text: "Hours of Operation", iconSvgText: null }}
      initiallyExpanded={false}
      id="hours"
    >
      <div
        className="m-schedule-page__sidebar-hours"
        // eslint-disable-next-line react/no-danger
        dangerouslySetInnerHTML={{ __html: hours }}
      />
    </ExpandableBlock>
  ) : null;

export default DefaultHoursOfOperation;
