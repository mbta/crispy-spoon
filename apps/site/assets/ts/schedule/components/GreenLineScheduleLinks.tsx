import React, { ReactElement } from "react";
import ExpandableBlock from "../../components/ExpandableBlock";
import renderFa from "../../helpers/render-fa";
import { SchedulePDF } from "./__schedule";

const pdfLink = (pdf: SchedulePDF): ReactElement<HTMLElement> => (
  <a key={pdf.url} href={pdf.url} rel="noopener noreferrer" target="_blank">
    Open subway schedule PDF {renderFa("fa-arrow-up-right-from-square", "")}
  </a>
);

const getGreenLineScheduleLinks = (): ReactElement<HTMLElement> => {
  return (
    <>
      <div>
        <a href="/schedules/Green-B/line">B Line Schedule</a>
      </div>
      <div>
        <a href="/schedules/Green-C/line">C Line Schedule</a>
      </div>
      <div>
        <a href="/schedules/Green-D/line">D Line Schedule</a>
      </div>
      <div>
        <a href="/schedules/Green-E/line">E Line Schedule</a>
      </div>
    </>
  );
};

const GreenLineScheduleLinks = ({
  pdfs
}: {
  pdfs: SchedulePDF[];
}): ReactElement<HTMLElement> | null => {
  console.log(pdfs);
  return (
    <>
      <ExpandableBlock
        header={{ text: "Weekday Schedule", iconSvgText: null }}
        initiallyExpanded={false}
        id="weekday-hours"
      >
        <div className="m-schedule-page__sidebar-hours">
          <div className="fs-14 pt-14">Trains every 6-12 minutes</div>
          {getGreenLineScheduleLinks()}
        </div>
      </ExpandableBlock>
      <ExpandableBlock
        header={{ text: "Weekend Schedule", iconSvgText: null }}
        initiallyExpanded={false}
        id="weekend-hours"
      >
        <div className="m-schedule-page__sidebar-hours">
          <div className="font-weight-bold fs-18 pb-14">Saturday</div>
          <div className="fs-14 pt-14">Trains every 6-12 minutes</div>
          {getGreenLineScheduleLinks()}
          <div
            style={{
              borderBottomWidth: "0.5px",
              borderBottomColor: "black",
              borderBottomStyle: "solid"
            }}
            className="pt-18"
          ></div>
          <div className="font-weight-bold fs-18 pt-18 pb-14">Sunday</div>
          {getGreenLineScheduleLinks()}
          {<div className="fs-14 pt-18">{pdfLink(pdfs[0])}</div>}
        </div>
      </ExpandableBlock>
    </>
  );
};

export default GreenLineScheduleLinks;
