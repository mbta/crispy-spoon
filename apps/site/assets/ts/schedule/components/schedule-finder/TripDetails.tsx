import React, { ReactElement } from "react";
import TripStop from "./TripStop";
import { TripInfo } from "../__trips";

export interface State {
  data: TripInfo | null;
  isLoading: boolean;
  error: boolean;
}

interface Props {
  state: State;
  showFare: boolean;
}

const TripSummary = ({
  tripInfo
}: {
  tripInfo: TripInfo;
}): ReactElement<HTMLElement> => (
  <tr>
    <td colSpan={3}>
      <div className="schedule-table__subtable-trip-info">
        <div className="schedule-table__subtable-trip-info-title u-small-caps u-bold">
          Trip length
        </div>
        {tripInfo.times.length} stops, {tripInfo.duration} minutes total
      </div>
      <div className="schedule-table__subtable-trip-info">
        <div className="schedule-table__subtable-trip-info-title u-small-caps u-bold">
          Fare
        </div>
        {tripInfo.fare && tripInfo.fare.price}
        <a
          className="schedule-table__subtable-trip-info-link"
          href={tripInfo.fare.fare_link}
        >
          View fares
        </a>
      </div>
    </td>
  </tr>
);

const allTimesHaveSchedule = (tripInfo: TripInfo): boolean =>
  tripInfo.times.every(time => !!time.schedule);

export const TripDetails = ({
  state,
  showFare
}: Props): ReactElement<HTMLElement> | null => {
  const { data: tripInfo, error, isLoading } = state;

  if (isLoading) {
    return (
      <div className="c-spinner__container">
        <div className="c-spinner">Loading...</div>
      </div>
    );
  }

  const errorLoadingTrip = (
    <p>
      <em>Error loading trip details. Please try again later.</em>
    </p>
  );

  if (error) {
    return errorLoadingTrip;
  }

  if (!tripInfo) return null;

  return (
    <table className="schedule-table__subtable">
      <thead>
        <TripSummary tripInfo={tripInfo} />
        <tr>
          <th scope="col" className="schedule-table__subtable-data">
            Stops
          </th>
          {showFare && (
            <th
              scope="col"
              className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
            >
              Fare
            </th>
          )}
          <th
            scope="col"
            className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
          >
            Arrival
          </th>
        </tr>
      </thead>
      <tbody className="schedule-table__subtable-tbody">
        {allTimesHaveSchedule(tripInfo)
          ? tripInfo.times.map((departure, index: number) => (
              <TripStop
                departure={departure}
                index={index}
                showFare={showFare}
                routeType={tripInfo.route_type}
                key={departure.schedule.stop.id}
              />
            ))
          : errorLoadingTrip}
      </tbody>
    </table>
  );
};
