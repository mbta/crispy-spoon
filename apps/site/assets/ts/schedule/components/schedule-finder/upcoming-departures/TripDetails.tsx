import React, { ReactElement } from "react";
import {
  TripDeparture,
  TripDepartureWithPrediction,
  TripInfo
} from "../../__trips";
import CrowdingPill from "../../line-diagram/CrowdingPill";
import TripStop from "./TripStop";

interface Props {
  tripInfo: TripInfo | null;
  showFare: boolean;
}

const hasAPrediction = (
  departure: TripDeparture
): departure is TripDepartureWithPrediction =>
  departure.prediction !== null &&
  departure.prediction.schedule_relationship !== "skipped";

const departuresWithPredictions = (
  departures: TripDeparture[]
): TripDepartureWithPrediction[] => departures.filter(hasAPrediction);

const TripSummary = ({
  tripInfo
}: {
  tripInfo: TripInfo;
}): ReactElement<HTMLElement> => (
  <tr className="trip-details-table__summary">
    <td colSpan={3} className="schedule-table__cell">
      <div>
        <span className="trip-details-table__title u-small-caps u-bold">
          Trip length
        </span>
        {departuresWithPredictions(tripInfo.times).length} stops,{" "}
        {tripInfo.duration} minutes total
      </div>
      <div>
        <span className="trip-details-table__title u-small-caps u-bold">
          Fare
        </span>

        {tripInfo.fare && tripInfo.fare.price}
        <a className="trip-details-table__link" href={tripInfo.fare.fare_link}>
          View fares
        </a>
      </div>
    </td>
  </tr>
);

const TripDetails = ({
  tripInfo,
  showFare
}: Props): ReactElement<HTMLElement> | null => {
  if (!tripInfo) return null;

  const crowding = tripInfo.vehicle?.crowding || null;

  return (
    <table className="trip-details-table">
      <thead>
        {crowding && (
          <tr>
            <th>
              <CrowdingPill crowding={crowding} />
            </th>
          </tr>
        )}
        <TripSummary tripInfo={tripInfo} />
        <tr>
          <th scope="col" className="schedule-table__cell">
            Stops
          </th>
          {showFare && (
            <th
              scope="col"
              className="schedule-table__cell schedule-table__cell--right-adjusted"
            >
              Fare
            </th>
          )}
          <th
            scope="col"
            className="schedule-table__cell schedule-table__cell--right-adjusted"
          >
            Arrival
          </th>
        </tr>
      </thead>
      <tbody>
        {departuresWithPredictions(tripInfo.times).map(
          (departure, index: number) => (
            <TripStop
              departure={departure}
              index={index}
              showFare={showFare}
              routeType={tripInfo.route_type}
              key={
                (departure.prediction.stop && departure.prediction.stop.id) ||
                departure.schedule.stop.id
              }
            />
          )
        )}
      </tbody>
    </table>
  );
};

export default TripDetails;
