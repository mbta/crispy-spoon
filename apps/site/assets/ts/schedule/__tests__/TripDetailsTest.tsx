import React from "react";
import renderer from "react-test-renderer";
import { createReactRoot } from "../../app/helpers/testUtils";
import { TripDetails, State } from "../components/schedule-finder/TripDetails";
import { TripInfo } from "../components/__trips";
import tripData from "./tripInfo.json";
import crTripData from "./crTripInfo.json";

const tripInfo: TripInfo = (tripData as unknown) as TripInfo;
const crTripInfo: TripInfo = (crTripData as unknown) as TripInfo;

const successState = {
  data: tripInfo,
  isLoading: false,
  error: false
} as State;

const crSuccessState = {
  data: crTripInfo,
  isLoading: false,
  error: false
} as State;

const errorState = {
  data: null,
  isLoading: false,
  error: true
} as State;

describe("TripDetails", () => {
  it("it renders trip details for a bus trip", () => {
    createReactRoot();
    const tree = renderer.create(
      <TripDetails state={successState} showFare={false} />
    );
    expect(tree).toMatchSnapshot();
  });

  it("it renders trip details for a CR trip", () => {
    createReactRoot();
    const tree = renderer.create(
      <TripDetails state={crSuccessState} showFare={true} />
    );
    expect(tree).toMatchSnapshot();
  });

  it("it renders an error if fetch failed", () => {
    createReactRoot();
    const tree = renderer.create(
      <TripDetails state={errorState} showFare={false} />
    );
    expect(tree).toMatchSnapshot();
  });
});
