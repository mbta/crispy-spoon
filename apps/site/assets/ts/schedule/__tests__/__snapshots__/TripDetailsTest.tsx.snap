// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TripDetails it renders an error if fetch failed 1`] = `
<p>
  <em>
    Error loading trip details. Please try again later.
  </em>
</p>
`;

exports[`TripDetails it renders trip details for a CR trip 1`] = `
<table
  className="schedule-table__subtable"
>
  <thead>
    <tr>
      <td
        colSpan={3}
      >
        <div
          className="schedule-table__subtable-trip-info"
        >
          <div
            className="schedule-table__subtable-trip-info-title u-small-caps u-bold"
          >
            Trip length
          </div>
          10
           stops, 
          69
           minutes total
        </div>
        <div
          className="schedule-table__subtable-trip-info"
        >
          <div
            className="schedule-table__subtable-trip-info-title u-small-caps u-bold"
          >
            Fare
          </div>
          $12.25
          <a
            className="schedule-table__subtable-trip-info-link"
            href="/fares/commuter_rail?destination=place-NEC-1851&origin=place-sstat"
          >
            View fares
          </a>
        </div>
      </td>
    </tr>
    <tr>
      <th
        className="schedule-table__subtable-data"
        scope="col"
      >
        Stops
      </th>
      <th
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
        scope="col"
      >
        Fare
      </th>
      <th
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
        scope="col"
      >
        Arrival
      </th>
    </tr>
  </thead>
  <tbody
    className="schedule-table__subtable-tbody"
  >
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-sstat"
        >
          South Station
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        02:30 PM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-bbsta"
        >
          Back Bay
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        $2.40
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        02:35 PM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-rugg"
        >
          Ruggles
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        $2.40
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        02:38 PM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-NEC-2173"
        >
          Route 128
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        $7.00
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        02:51 PM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-NEC-2139"
        >
          Canton Junction
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        $8.00
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        02:57 PM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-NEC-2108"
        >
          Sharon
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        $8.75
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        03:03 PM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-NEC-2040"
        >
          Mansfield
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        $10.50
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        03:11 PM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-NEC-1969"
        >
          Attleboro
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        $11.00
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        03:19 PM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-NEC-1919"
        >
          South Attleboro
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        $11.00
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        03:29 PM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-NEC-1851"
        >
          Providence
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        $12.25
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        03:39 PM
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TripDetails it renders trip details for a bus trip 1`] = `
<table
  className="schedule-table__subtable"
>
  <thead>
    <tr>
      <td
        colSpan={3}
      >
        <div
          className="schedule-table__subtable-trip-info"
        >
          <div
            className="schedule-table__subtable-trip-info-title u-small-caps u-bold"
          >
            Trip length
          </div>
          11
           stops, 
          14
           minutes total
        </div>
        <div
          className="schedule-table__subtable-trip-info"
        >
          <div
            className="schedule-table__subtable-trip-info-title u-small-caps u-bold"
          >
            Fare
          </div>
          $2.40
          <a
            className="schedule-table__subtable-trip-info-link"
            href="/fares/bus-fares"
          >
            View fares
          </a>
        </div>
      </td>
    </tr>
    <tr>
      <th
        className="schedule-table__subtable-data"
        scope="col"
      >
        Stops
      </th>
      <th
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
        scope="col"
      >
        Arrival
      </th>
    </tr>
  </thead>
  <tbody
    className="schedule-table__subtable-tbody"
  >
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-sstat"
        >
          South Station
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        05:46 AM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-crtst"
        >
          Courthouse
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        05:48 AM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-wtcst"
        >
          World Trade Center
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        05:50 AM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/place-conrd"
        >
          Silver Line Way
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        05:53 AM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/247"
        >
          Northern Ave @ Harbor St
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        05:54 AM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/30249"
        >
          Northern Ave @ Tide St
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        05:55 AM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/30250"
        >
          23 Drydock Ave
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        05:56 AM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/30251"
        >
          27 Drydock Ave
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        05:56 AM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/31259"
        >
          Drydock Ave @ Black Falcon Ave
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        05:56 AM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/31258"
        >
          88 Black Falcon
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        05:57 AM
      </td>
    </tr>
    <tr
      className="schedule-table__subtable-row"
    >
      <td
        className="schedule-table__subtable-data"
      >
        <a
          href="/stops/31255"
        >
          Drydock Ave @ Design Center Place
        </a>
      </td>
      <td
        className="schedule-table__subtable-data schedule-table__subtable-data--right-adjusted"
      >
        06:00 AM
      </td>
    </tr>
  </tbody>
</table>
`;
